src/Symfony/Bundle/WebProfilerBundle/Csp/ContentSecurityPolicyHandler.php

```diff
@@ -0,0 +1,272 @@
+
+/**
+ * Handles Content-Security-Policy HTTP header for the WebProfiler Bundle.
+ *
+ * @author Romain Neutron <imprec@gmail.com>
+ */
+class ContentSecurityPolicyHandler
+{
+    private $nonceGenerator;
+    private $cspForceDisabled = false;
+
+    public function __construct(NonceGenerator $nonceGenerator)
+    {
+        return $this->nonceGenerator = $nonceGenerator;
+    }
+
+    /**
+     * Returns an array of nonces to be used in Twig templates and Content-Security-Policy headers.
+     *
+     * Nonce can be provided by;
+     *  - The request - In case HTML content is fetched via AJAX and inserted in DOM, it must use the same nonce as origin
+     *  - The response -  A call to getNonces() has already been done previously. Same nonce are returned
+     *  - They are otherwise randomly generated
+     *
+     * @param Request  $request
+     * @param Response $response
+     *
+     * @return array
+     */
+    public function getNonces(Request $request, Response $response)
+    {
+        if ($request->headers->has('X-SymfonyProfiler-Script-Nonce') && $request->headers->has('X-SymfonyProfiler-Style-Nonce')) {
+            return array(
+                'csp_script_nonce' => $request->headers->get('X-SymfonyProfiler-Script-Nonce'),
+                'csp_style_nonce' => $request->headers->get('X-SymfonyProfiler-Style-Nonce'),
+            );
+        }
+
+        if ($response->headers->has('X-SymfonyProfiler-Script-Nonce') && $response->headers->has('X-SymfonyProfiler-Style-Nonce')) {
+            return array(
+                'csp_script_nonce' => $response->headers->get('X-SymfonyProfiler-Script-Nonce'),
+                'csp_style_nonce' => $response->headers->get('X-SymfonyProfiler-Style-Nonce'),
+            );
+        }
+
+        $nonces = array(
+            'csp_script_nonce' => $this->generateNonce(),
+            'csp_style_nonce' => $this->generateNonce(),
+        );
+
+        $response->headers->set('X-SymfonyProfiler-Script-Nonce', $nonces['csp_script_nonce']);
+        $response->headers->set('X-SymfonyProfiler-Style-Nonce', $nonces['csp_style_nonce']);
+
+        return $nonces;
+    }
+
+    public function forceDisableCsp()
+    {
+        $this->cspForceDisabled = true;
+    }
+
+    /**
+     * Cleanup temporary headers and updates Content-Security-Policy headers.
+     *
+     * This method should be called on KernelEvents::RESPONSE
+     *
+     * @param Request  $request
+     * @param Response $response
+     *
+     * @return array Nonce used by the bundle in Content-Security-Policy header
+     */
+    public function onKernelResponse(Request $request, Response $response)
+    {
+        if ($this->cspForceDisabled) {
+            $this->removeCspHeaders($response);
+
+            return array();
+        }
+
+        $nonces = $this->getNonces($request, $response);
+        $this->cleanHeaders($response);
+        $this->updateCspHeaders($response, $nonces);
+
+        return $nonces;
```

Comment:
OK, this method is actually not an event listener. So the phpdoc is wrong (and the naming is not that good either)